File: common/src/main/java/generations/gg/generations/core/generationscore/client/render/rarecandy/ModelRegistry.java
package generations.gg.generations.core.generationscore.client.render.rarecandy;

import com.google.common.cache.*;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.mojang.blaze3d.pipeline.RenderCall;
import com.mojang.blaze3d.systems.RenderSystem;
import com.mojang.blaze3d.vertex.PoseStack;
import com.mojang.math.Axis;
import generations.gg.generations.core.generationscore.GenerationsCore;
import generations.gg.generations.core.generationscore.client.model.ModelContextProviders;
import generations.gg.generations.core.generationscore.world.level.block.entities.ModelProvidingBlockEntity;
import generations.gg.generations.core.generationscore.world.level.block.generic.GenericRotatableModelBlock;

import gg.generations.rarecandy.renderer.animation.Animation;
import gg.generations.rarecandy.renderer.components.AnimatedMeshObject;
import gg.generations.rarecandy.renderer.components.MeshObject;
import gg.generations.rarecandy.renderer.rendering.RareCandy;
import net.minecraft.client.Minecraft;
import net.minecraft.resources.ResourceLocation;
import org.jetbrains.annotations.NotNull;

import java.io.FileNotFoundException;
import java.util.AbstractMap;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.TimeUnit;
import java.util.function.Consumer;
import java.util.function.Supplier;

import static com.google.common.util.concurrent.Futures.immediateFuture;

public class ModelRegistry {
    private static final String DUMMY = "dummy";

    public static final LoadingCache<ResourceLocation, CompiledModel> LOADER = CacheBuilder.newBuilder().expireAfterAccess(10, TimeUnit.SECONDS).removalListener((RemovalListener<ResourceLocation, CompiledModel>) notification -> notification.getValue().delete()).build(new CacheLoader<>() {
        @Override
        public @NotNull CompiledModel load(@NotNull ResourceLocation pair) {
            var resourceManager = Minecraft.getInstance().getResourceManager();

            try {
                return CompiledModel.of(pair, resourceManager.getResourceOrThrow(pair));
            } catch (FileNotFoundException e) {
                throw new RuntimeException(e);
            }
        }
    });
    private static RareCandy WORLD_RENDER;
    private static RareCandy GUI_RENDER;

    public static CompiledModel get(ModelContextProviders.ModelProvider modelProvider) {
        return get(modelProvider.getModel());
    }

    public static CompiledModel get(ResourceLocation location) {
        try {
            return LOADER.get(location);

        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    public static void prepForBER(PoseStack stack, ModelContextProviders.AngleProvider supplier) {

        stack.translate(0.5f, 0.0f, 0.5f);

        if(supplier instanceof ModelProvidingBlockEntity blockEntity && blockEntity.getBlockState().getBlock() instanceof GenericRotatableModelBlock<?> block && block.shouldRotateSpecial()) {
            var forward = blockEntity.getBlockState().getValue(GenericRotatableModelBlock.FACING);

            var x = block.getWidthValue(blockEntity.getBlockState());
            var z = block.getLengthValue(blockEntity.getBlockState());


            var width = (block.width() * 0.5f) - x;
            var length = (block.length() * 0.5f) - z;

            switch (forward) {
                case SOUTH -> stack.translate(width, 0, -length);
                case EAST -> stack.translate(-length, 0, -width);
                case NORTH -> stack.translate(-width, 0, length);
                case WEST -> stack.translate(length, 0, width);
            }

            stack.mulPose(Axis.YN.rotationDegrees(supplier.getAngle()));
        } else {
           stack.mulPose(Axis.YN.rotationDegrees(supplier.getAngle()));
        }

    }

    public static RareCandy getWorldRareCandy() {
        //RareCandy.DEBUG_THREADS = true;
        if (WORLD_RENDER == null) WORLD_RENDER = new RareCandy();
        Animation.animationModifier = (animation, s) -> {
            if (s.equals("gfb")) animation.ticksPerSecond = 60_000; // 60 fps. 1000 ticks per frame?
        };
        return WORLD_RENDER;
    }

    public static RareCandy getGuiRareCandy() {
        //RareCandy.DEBUG_THREADS = true;
        if (GUI_RENDER == null) GUI_RENDER = new RareCandy();
        Animation.animationModifier = (animation, s) -> {
            if (s.equals("gfb")) animation.ticksPerSecond = 60_000; // 60 fps. 1000 ticks per frame?
        };
        return GUI_RENDER;
    }
}

File: common/src/main/java/generations/gg/generations/core/generationscore/client/render/rarecandy/ModelRegistry.kt
package generations.gg.generations.core.generationscore.client.render.rarecandy

import com.github.benmanes.caffeine.cache.CacheLoader
import com.github.benmanes.caffeine.cache.Caffeine
import com.github.benmanes.caffeine.cache.RemovalCause
import com.github.benmanes.caffeine.cache.RemovalListener
import com.mojang.blaze3d.vertex.PoseStack
import com.mojang.math.Axis
import generations.gg.generations.core.generationscore.client.model.ModelContextProviders
import generations.gg.generations.core.generationscore.client.model.ModelContextProviders.AngleProvider
import generations.gg.generations.core.generationscore.client.render.rarecandy.CompiledModel.Companion.of
import generations.gg.generations.core.generationscore.world.level.block.entities.ModelProvidingBlockEntity
import generations.gg.generations.core.generationscore.world.level.block.generic.GenericRotatableModelBlock
import gg.generations.rarecandy.renderer.animation.Animation
import gg.generations.rarecandy.renderer.rendering.RareCandy
import net.minecraft.client.Minecraft
import net.minecraft.core.Direction
import net.minecraft.resources.ResourceLocation
import java.io.FileNotFoundException
import java.util.concurrent.TimeUnit
import java.util.function.BiConsumer

object ModelRegistry {
    private const val DUMMY = "dummy"
    private val LOADER = Caffeine.newBuilder().expireAfterAccess(1, TimeUnit.HOURS).removalListener { _: ResourceLocation?, value: CompiledModel?, _: RemovalCause? -> value!!.delete() }
        .buildAsync<ResourceLocation, CompiledModel?>(
            CacheLoader { key ->
                val resourceManager = Minecraft.getInstance().resourceManager
                try {
                    return@CacheLoader of(key!!, resourceManager.getResourceOrThrow(key))
                } catch (e: FileNotFoundException) {
                    throw RuntimeException(e)
                }
            })
    @JvmStatic
    operator fun get(modelProvider: ModelContextProviders.ModelProvider): CompiledModel? {
        return ModelRegistry[modelProvider.model]
    }

    @JvmStatic
    fun clear() = LOADER.asMap().clear();

    @JvmStatic
    operator fun get(location: ResourceLocation): CompiledModel? {
        return try {
            LOADER[location].getNow(null)
        } catch (e: Exception) {
            throw RuntimeException(e)
        }
    }

    @JvmStatic
    fun prepForBER(stack: PoseStack, supplier: AngleProvider) {
        stack.translate(0.5f, 0.0f, 0.5f)
        if (supplier is ModelProvidingBlockEntity) {
            val block = supplier.blockState.block;

            if (block is GenericRotatableModelBlock<*> && block.shouldRotateSpecial()) {
                val forward: Direction = supplier.blockState.getValue(GenericRotatableModelBlock.FACING)
                val x: Int = block.getWidthValue(supplier.blockState)
                val z: Int = block.getLengthValue(supplier.blockState)
                val width: Float = block.width() * 0.5f - x
                val length: Float = block.length() * 0.5f - z
                when (forward) {
                    Direction.SOUTH -> stack.translate(width, 0f, -length)
                    Direction.EAST -> stack.translate(-length, 0f, -width)
                    Direction.NORTH -> stack.translate(-width, 0f, length)
                    Direction.WEST -> stack.translate(length, 0f, width)
                    else -> {}
                }
                stack.mulPose(Axis.YN.rotationDegrees(supplier.getAngle()))
            } else {
                stack.mulPose(Axis.YN.rotationDegrees(supplier.angle))
            }
        } else {
            stack.mulPose(Axis.YN.rotationDegrees(supplier.angle))
        }
    }

    @JvmStatic
    val worldRareCandy: RareCandy = RareCandy().also {
        Animation.animationModifier = BiConsumer { animation: Animation, s: String ->
            if (s == "gfb") animation.ticksPerSecond = 60000f // 60 fps. 1000 ticks per frame?
        }
    }
}